gnuradio-runtime/lib/prefs.cc

```diff
@@ -72,62 +77,58 @@ void prefs::_read_files(const std::vector<std::string>& filenames)
-                    // value of a basic_option is always a std::vector<string>; we only
-                    // allow single values, so:
-                    std::string value = o.value[0];
-                    d_config_map[section][key] = value;
+        if (!infile.good()) {
+            std::cerr << "WARNING: Config file '" << fname
+                      << "' could not be opened for reading." << std::endl;
+            continue;
+        }
+
+        try {
+            po::basic_parsed_options<char_t> parsed =
+                po::parse_config_file(infile, po::options_description(), true);
+            for (const auto& o : parsed.options) {
+                std::string okey = o.string_key;
+                size_t pos = okey.find(".");
+                std::string section, key;
+                if (pos != std::string::npos) {
+                    section = okey.substr(0, pos);
+                    key = okey.substr(pos + 1);
+                } else {
+                    section = "default";
+                    key = okey;
                 }
-            } catch (std::exception& e) {
-                std::cerr << "WARNING: Config file '" << fname
-                          << "' failed to parse:" << std::endl;
-                std::cerr << e.what() << std::endl;
-                std::cerr << "Skipping it" << std::endl;
+                stolower(section);
+                stolower(key);
+                // value of a basic_option is always a std::vector<string>; we only
+                // allow single values, so:
+                std::string value = o.value[0];
+                std::lock_guard<std::mutex> lk(d_mutex);
+                d_config_map[section][key] = value;
             }
-        } else { // infile.good();
+        } catch (std::exception& e) {
             std::cerr << "WARNING: Config file '" << fname
-                      << "' could not be opened for reading." << std::endl;
+                      << "' failed to parse:" << std::endl;
+            std::cerr << e.what() << std::endl;
+            std::cerr << "Skipping it" << std::endl;
         }
     }
 }
 
 void prefs::add_config_file(const std::string& configfile)
 {
-    std::vector<std::string> filenames;
-    filenames.push_back(configfile);
-
-    _read_files(filenames);
+    _read_files(std::vector<std::string>{ configfile });
 }
 
 
 std::string prefs::to_string()
 {
-    config_map_itr sections;
-    config_map_elem_itr options;
     std::stringstream s;
+    std::lock_guard<std::mutex> lk(d_mutex);
 
-    for (sections = d_config_map.begin(); sections != d_config_map.end(); sections++) {
-        s << "[" << sections->first << "]" << std::endl;
-        for (options = sections->second.begin(); options != sections->second.end();
-             options++) {
-            s << options->first << " = " << options->second << std::endl;
+    for (const auto& sections : d_config_map) {
+        s << "[" << sections.first << "]" << std::endl;
```

Comment:
Well, we know this is not wchar, and we don't need any flushing behaviour, so generally, this should be "\n" instead of std::endl, but let's not touch our output routines in this PR.